Updated Modifications to obs setup code(setup*.f90) for ps/w/spd/q/t to add observation provider info into obs diagnostic file

GangZhao-NOAA opened the pull request:
Adding modifications to obs setup code(setup*.f90) for ps/q/t/w/spd, to include the observation provider and sub-provider information into the obsdiags filee. Adding new option "l_obsprvdiag" to trigger the output.   These provider information can be read with read_diag_conv and used in the AutoObsQC for 3DRTMA.

Also modifying the code in read_diag_conv.f90 under util/Analysis_Utilities/read_diag to read the provider information in the obs-diag file and write out in text format which can be used for auto-obsqc.

GangZhao-NOAA committed:
Adding modifications to obs setup code(setup*.f90) for
ps/q/t/w/spd, to include the observation provider and
sub-provider information into the diag files, which
will be used in read_diag_conv_3drtma in Auto-ObsQC for
3DRTMA.

Also adding a new spcific code read_diag_conv_3drtma.f90
under read_diag to read the osb-diag file which includes
the obs provider information for auto-qc.

GangZhao-NOAA committed:
Adding option l_obsprvdiag (instead of l_rtma3d) to control the
observation provider information output into the obsdiags files. Also
adding the code in read_diag_conv.f090 to read the obs provider info
from obsdiags file, then read_diag_conv_3drtma.f90 is no longer
necessary and therefore removed from the code tree.

GangZhao-NOAA committed:
Merge branch 'NOAA-GSL:feature/rrfs' into feature/rrfs

guoqing-noaa commented on src/gsi/gsimod.F90:
```diff
@@ -712,7 +719,8 @@ module gsimod
        npePgrp_rfv3,rfv3_pe_T,rfv3_pe_v,rfv3_pe_q,rfv3_pe_ps,rfv3_pe_dz, &
-       l_reg_update_hydro_delz
+       l_reg_update_hydro_delz, &
+       l_obsprvdiag
```
Line 723 can be put into line 722.

GangZhao-NOAA replied:
I will fix it.

guoqing-noaa commented on src/gsi/gsimod.F90:
```diff
@@ -1757,6 +1765,9 @@ subroutine gsimain_initialize
 
+! Ensure when running 2DVAR (RTMA) or 3DRTMA, writing out obs provider info
+  if (twodvar_regional .or. l_rtma3d) l_obsprvdiag = .true.
+
```
It is better NOT to force (or hardcoded) this inside the code as it hides information from general users.
For 2D/3D RTMA, setting l_obsprvdiag = .true. in the namelist is enough and provides a flexibility/backward compatibility to NOT to output subprovider information.

GangZhao-NOAA replied:
I will remove this line.

guoqing-noaa commented on src/gsi/setupps.f90:
```diff
@@ -321,7 +328,10 @@ subroutine setupps(obsLL,odiagLL,lunin,mype,bwork,awork,nele,nobs,is,conv_diagsa
      nreal=ioff0
      if (lobsdiagsave) nreal=nreal+4*miter+1
-     if (twodvar_regional) then; nreal=nreal+2; allocate(cprvstg(nobs),csprvstg(nobs)); endif
+     if (twodvar_regional .or. l_rtma3d .or. l_obsprvdiag) then
```
Line 331 only needs `if (l_obsprvdiag) then`

As discussed before, let "l_obsprvdiag" solely decide what to do, not if it is 2D/3D RTMA.  If RTMA wants to output subprovider information, just set "l_obsprvdiag" to "true" in the namelist. 

Suggest to remove all these kind of unnecessary if (twodvar_regional .or. l_rtma3d in other places.

Also note that "l_rtma3d" is previously added for ghost reflectivity removal in the cloud analysis part. If a RTMA version does not run cloud analysis, using "if l_rtma3d" will be problematic.

GangZhao-NOAA replied:
I can remove the trigger of "l_rtma3d" in setup?.f90 to control the output of obs provider info. 
twodvar_regional might still be necessary for backward compatible to operational (2D)RTMA sono need to modify the script and namelist file to add the "l_obsprvdiag". Anyway, I will discuss with Manuel and Matthew for this.

GangZhao-NOAA commented:
@guoqing-noaa  @hu5970 

Thank you very much for your comments/suggestions on the code. I will discuss with Manuel and Matthew about the necessity of keeping "twodvar_regional" in setup??.f90. 
Today aftger discussion with Matthew, I found a few questions in the read_diag_conv.f90. Matthew is actually running the code without using namelist, but reading the parameters from command line (like read_diag_conv.x infilename outfilename true false). I am asking them which way (namelist or command argument list) they prefer to run.
I may have to take the PR back for update, then re-submit for review.

Thank you very much!

guoqing-noaa commented:
@GangZhao-NOAA Thanks for looking on this. You don't need to close this PR and then create a new one. Just push new changes to your feature/rrfs branch and it will be reflected here.

GangZhao-NOAA committed:
Minor modifcations corrsponding to code reviewer's suggestions:
1. removal of l_rtma3d in setup*.f90 (for t/q/ps/w/spd), so only
l_obsprvdiag or twodvar_regional is used to trigger the output of obs
provider info into obsdiags file;

guoqing-noaa commented on src/gsi/setupq.f90:
```diff
@@ -394,7 +400,11 @@ subroutine setupq(obsLL,odiagLL,lunin,mype,bwork,awork,nele,nobs,is,conv_diagsav
+       nreal=nreal+2                            ! account for idomsfc,izz
+       allocate(cprvstg(nobs),csprvstg(nobs))   ! obs provider info
+       if(l_pbl_pseudo_surfobsq) allocate(cprvstgp(nobs*3),csprvstgp(nobs*3)) ! obs provider info for pseudo obs
+       endif
```
remove two spaces before line 407 so that it aligns with line 403

GangZhao-NOAA replied:
@guoqing-noaa @hu5970
Thank you for the review. 
The two spaces in beginning of the line 407 of "setupq/f90" was removed. I make a new commit and push it back. 

As for comparing with old code for validation, I have not run it yet.  My colleague Matthew Morris is going to test it. I will let you know the results of comparison. But he will only test with HRRR-based run, not for FV3-LAM based rrfs run.

Thank you!

guoqing-noaa reviewed (approved):
Looks good to me (with one minor formatting comment). Thanks!

Another thing, have you run comparison tests to make sure this PR dose not change the GSI analysis result?

GangZhao-NOAA committed:
minor format modification in gsi/setupq.f90

GangZhao-NOAA commented:
@guoqing-noaa @hu5970 
Hi, Ming and Guoqing,

I found an issue with the "read_diag_conv.f90" and need to discuss with Manuel and Matthew about the way to fix it. Please do not merge the code for now. 
Thank you very much! Happy Holiday!

- Gang

GangZhao-NOAA committed:
A bug (missing the reset of iflg for provider) was fixed in
read_diag_cov.f90.

GangZhao-NOAA committed:
fixing bugs: both work when need to read/write provider info
1. not re-set iflg when checking up cprovider
2. a bug in the if-block to write out the obs diag info into a text
format file.

GangZhao-NOAA commented:
@hu5970 @guoqing-noaa 
Hi Guoqing and Ming,

I fixed the bugs in read_diag_conv.f90 (see the latest [commit # 295fdf](https://github.com/NOAA-GSL/GSI/pull/8/commits/e295fdf412a814f9f7c0ca0639e865dc300ab4a3)). Matthew Morris of EMC had tested it and confirm that it is correct now.

As for your request to compare the analysis results from my modified GSI code (feature/rrfs in my fork of GSI) and the old GSI code (feature/rrfs of NOAA-GSL GSI), I compiled both GSI package on WCOSS. Matthew made the test runs on WCOSS. The analysis results are identical. So I believe the modified code are good.

So the modified code should be ready for merging into NOAA-GSL/GSI. We hope the merging could be done soon, so we can work on the same code.

Thank you very much!

- Gang

GangZhao-NOAA commented:
@guoqing-noaa @hu5970 
Thank you so much!
